spirv-fuzz: Handle invalid ids in fact manager

Vasniktel opened the pull request:
Fixes #3741.

This PR fixes two bugs:
1. Segfault in https://github.com/KhronosGroup/SPIRV-Tools/blob/4c239bd81bf79337e4a0cfe1c9da9caf1b25108a/source/fuzz/fuzzer_util.cpp#L574-L576 when `result_id` doesn't exist in the module. The bug is triggered in https://github.com/KhronosGroup/SPIRV-Tools/blob/4c239bd81bf79337e4a0cfe1c9da9caf1b25108a/source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp#L290 when `representative->object()` doesn't exist in the module. This can happen in certain cases if we apply `TransformationMergeBlocks` first.
2. Segfault in either https://github.com/KhronosGroup/SPIRV-Tools/blob/4c239bd81bf79337e4a0cfe1c9da9caf1b25108a/source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp#L769 or https://github.com/KhronosGroup/SPIRV-Tools/blob/4c239bd81bf79337e4a0cfe1c9da9caf1b25108a/source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp#L772 happens if either `dd1.object()` or `dd2.object()` doesn't exist in the module. This can happen when `AddDataSynonymFactRecursive` is called recursively from `ComputeConversionDataSynonymFacts`. This bug is described in #3741.

Vasniktel commented on source/fuzz/fact_manager.cpp:
```diff
@@ -761,8 +756,13 @@ void FactManager::DataSynonymAndIdEquationFacts::
     std::vector<const protobufs::DataDescriptor*> convert_u_to_f_lhs;
 
     for (const auto& fact : id_equations_) {
+      if (!context->get_def_use_mgr()->GetDef(fact.first->object())) {
+        // Skip non-existent instructions.
+        continue;
+      }
```
This might be too extreme. We could mark two data descriptors as synonymous even if one of them doesn't exist in the module. We can do that if elements of equivalence classes of both data descriptors are guaranteed to have compatible types.

This can be useful since some of the elements of those equivalence classes might still exist in the module.

A drawback of this approach, though, is that it might be harder to manually analyze a shader in this case.

Anyway, we can't do this ATM since an assertion fails in
https://github.com/KhronosGroup/SPIRV-Tools/blob/4dd122392f3ad757e70951a1198479bf233d4cd8/source/fuzz/fact_manager.cpp#L720-L723

Vasniktel replied:
@afd, @paulthomson, do you think we should add synonyms between non-existent data descriptors?

afd reviewed (changes requested):
This is a really important issue for us to fix, and as it's a bit unclear how best to proceed I think it's really important we have a simple PR for the proposed fix, with a test case that exposes the problem.

I thus wonder whether you could:

(a) do a separate PR containing all the changes here that are about cleaning up data_synonym_and_id_equation_facts (those changes look good, but I wasn't sure which were related to the fix)

(b) turn this into a smaller PR that just contributes the fix, and also contributes a test case that exposes the problem.  But could the test case expose the problem by applying some transformations, not by killing an instruction?  For example I think you said that block merging can lead to OpPhi instructions disappearing; could you write a test based on that?

Vasniktel commented:
@afd, I've updated the description. Regarding the points mentioned in your previous comment:
1. All changes in the PR fix the issue.
2. I've updated the test from #3741. The new test uses both `TransformationMergeBlocks` and `KillDef` to test two bugs as mentioned in the description comment accordingly.

Let me know if you need any clarification.

afd reviewed (approved):
This looks good, thanks!
